File: java_solutions/103.binary-tree-zigzag-level-order-traversal.java
/*
 * @lc app=leetcode id=103 lang=java
 *
 * [103] Binary Tree Zigzag Level Order Traversal
 */

class Solution {
    public java.util.List<java.util.List<Integer>> zigzagLevelOrder(TreeNode root) {
        return root == null ? java.util.Collections.emptyList() : java.util.stream.Stream.iterate(new Object[]{java.util.Collections.singletonList(root), 0}, s -> !((java.util.List<TreeNode>)s[0]).isEmpty(), s -> new Object[]{((java.util.List<TreeNode>)s[0]).stream().flatMap(n -> java.util.stream.Stream.of(n.left, n.right).filter(java.util.Objects::nonNull)).collect(java.util.stream.Collectors.toList()), (int)s[1] + 1}).map(s -> java.util.stream.Stream.of(((java.util.List<TreeNode>)s[0]).stream().map(n -> n.val).collect(java.util.stream.Collectors.toList())).map(l -> (int)s[1] % 2 == 1 ? java.util.stream.Stream.of(l).peek(java.util.Collections::reverse).findFirst().get() : l).findFirst().get()).collect(java.util.stream.Collectors.toList());
    }
}

File: kotlin_solutions/103.binary-tree-zigzag-level-order-traversal.kt
/*
 * @lc app=leetcode id=103 lang=kotlin
 *
 * [103] Binary Tree Zigzag Level Order Traversal
 */

class Solution { fun zigzagLevelOrder(root: TreeNode?): List<List<Int>> = root?.let { generateSequence(listOf(it) to false) { (level, reverse) -> level.flatMap { listOfNotNull(it.left, it.right) }.takeIf { it.isNotEmpty() }?.let { it to !reverse } }.map { (level, reverse) -> level.map { it.`val` }.let { if (reverse) it.reversed() else it } }.toList() } ?: emptyList() }

